Update runtime model to support multiple resources per metric

adamconnelly opened the pull request:
Works towards #513 

## Proposed Changes

This PR updates the runtime model to support multiple resources per metric, but without altering the v1 config format. It shouldn't cause any change in functionality, but gives us the basis to update the config to support multiple resources per metric.

@tomkerkhove I've kept this split into multiple commits because it's a pretty huge change, and I figured it would be a bit easier to follow what I was trying to do. The first commit contains the main change, and then the other commits are really just the required updates to the rest of the codebase to support that change.

tomkerkhove commented:
I will have a look, thanks!

*PS: Build failure is due to Azure DevOps, next commit it will kick in again or will find another way to sort it out.*

tomkerkhove commented on src/Promitor.Core.Scraping/Configuration/Providers/MetricsDeclarationProvider.cs:
```diff
@@ -32,9 +32,12 @@ public virtual MetricsDeclaration Get(bool applyDefaults = false)
-                        metric.ResourceGroupName = config.AzureMetadata.ResourceGroupName;
+                        if (string.IsNullOrWhiteSpace(resource.ResourceGroupName))
+                        {
+                            resource.ResourceGroupName = config.AzureMetadata.ResourceGroupName;
```
Since we already do a check in the scraper - WHat happens if we leave this out? Or can we assign the correct rg to scrape here already and get rid of it in the scraper?

adamconnelly replied:
You're completely right - we don't actually need this because the logic is also in the Scraper. I've removed this and made the setter on `ResourceGroupName` private.

tomkerkhove commented on src/Promitor.Core.Scraping/Configuration/Serialization/v1/Core/ConfigurationSerializer.cs:
```diff
@@ -43,6 +44,7 @@ public MetricsDeclarationV1 InterpretYamlStream(YamlMappingNode rootNode)
 
             var metricsDeclaration = new MetricsDeclarationV1
             {
+                Version = SpecVersion.v1.ToString(),
```
Would it make sense to just move this to `MetricsDeclarationV1` itself?

adamconnelly replied:
Sure.

adamconnelly commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/AzureResourceDefinition.cs:
```diff
@@ -0,0 +1,30 @@
+        ///     Specify a resource group to scrape that defers from the default resource group.
+        ///     This enables you to do multi-resource group scraping with one configuration file.
+        /// </summary>
+        public string ResourceGroupName { get; set; }
```
Unfortunately we do because of https://github.com/tomkerkhove/promitor/blob/master/src/Promitor.Core.Scraping/Configuration/Providers/MetricsDeclarationProvider.cs#L37.

If we wanted to remove the setter I guess we'd need to rethink how the defaulting worked.

adamconnelly replied:
Just ignore that comment - as you pointed out the defaulting code in MetricsDeclarationProvider isn't actually required, so I was able to make this private.

tomkerkhove replied:
Awesome, thanks!

adamconnelly commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/PrometheusMetricDefinition.cs:
```diff
@@ -0,0 +1,29 @@
+{
+    public class PrometheusMetricDefinition
+    {
+        public PrometheusMetricDefinition(string name, string description, Dictionary<string, string> labels)
```
No problem.

adamconnelly commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/PrometheusMetricDefinition.cs:
```diff
@@ -0,0 +1,29 @@
+    public class PrometheusMetricDefinition
+    {
+        public PrometheusMetricDefinition(string name, string description, Dictionary<string, string> labels)
+        {
```
So this is slightly tricky - it goes back to the way the validation stuff works where it's currently done against the runtime model, and it expects a null name to be allowed (presumably so that we can display a slightly friendlier message on startup?).

If I add the guards in a load of the tests under `Validation.Metrics.ResourceTypes` start failing.

I could maybe try to change the validation code to work against the config model rather than the runtime model, but I wanna hold off until I hear your thoughts on it.

tomkerkhove replied:
I think the current approach is ok to work against the runtime model and validate that as that's what really matters, thanks for checking!

Resolving.

adamconnelly commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/ResourceTypes/ContainerRegistryResourceDefinition.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class ContainerRegistryResourceDefinition : AzureResourceDefinition
+    {
+        public ContainerRegistryResourceDefinition() : base(ResourceType.ContainerRegistry)
```
They actually aren't needed - I'd left them in because there was a `new()` type constraint on the `Scraper` class, but it doesn't look like that's required, so I've removed the constraint and the empty constructors now.

I had to add a custom mapping in for the `ServiceBusQueueMetricDefinitionV1 -> ServiceBusQueueResourceDefinition` mapping because the parameter name in the constructor is `ns` which doesn't match the property name, `Namespace`. Just for the record, the catch all test in the `AutoMapperTests` caught this, so I didn't add an extra test for it.

adamconnelly commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/ResourceTypes/ContainerRegistryResourceDefinition.cs:
```diff
@@ -0,0 +1,17 @@
+            RegistryName = registryName;
+        }
+
+        public string RegistryName { get; set; }
```
No problem.

adamconnelly commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/ScrapeDefinition.cs:
```diff
@@ -0,0 +1,37 @@
+        /// </summary>
+        public AzureMetricConfiguration AzureMetricConfiguration { get; }
+
+        public PrometheusMetricDefinition PrometheusMetricDefinition { get; }
```
Sure.

adamconnelly commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/ScrapeDefinition.cs:
```diff
@@ -0,0 +1,37 @@
+            PrometheusMetricDefinition prometheusMetricDefinition,
+            Scraping scraping,
+            TResourceDefinition resource)
+        {
```
Sure.

adamconnelly commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/ScrapeDefinition.cs:
```diff
@@ -0,0 +1,37 @@
+    /// </summary>
+    public class ScrapeDefinition<TResourceDefinition> where TResourceDefinition: AzureResourceDefinition
+    {
+        public ScrapeDefinition(
```
No problem.

adamconnelly commented on src/Promitor.Core.Scraping/Configuration/Model/MetricsDeclaration.cs:
```diff
@@ -5,7 +5,6 @@ namespace Promitor.Core.Scraping.Configuration.Model
 {
     public class MetricsDeclaration
     {
-        public string Version { get; set; }
```
Yeah - it wouldn't have caused any issue to leave it, but I figured that it didn't really make sense since the internal model doesn't have a version - it's just the configuration that is versioned.

adamconnelly commented on src/Promitor.Scraper.Host/Validation/MetricDefinitions/MetricsValidator.cs:
```diff
@@ -44,7 +44,7 @@ private IEnumerable<string> Validate(MetricDefinition metric)
             }
 
-            if (string.IsNullOrWhiteSpace(metric.Name))
+            if (string.IsNullOrWhiteSpace(metric.PrometheusMetricDefinition.Name))
```
Sure - I've altered it to use null propagation.

adamconnelly commented on src/Promitor.Scraper.Host/Validation/MetricDefinitions/ResourceTypes/ContainerInstanceMetricValidator.cs:
```diff
@@ -1,18 +1,24 @@
+            Guard.NotNull(metricDefinition, nameof(metricDefinition));
 
-            if (string.IsNullOrWhiteSpace(containerInstanceMetricDefinition.ContainerGroup))
+            foreach (var definition in metricDefinition.Resources.Cast<ContainerInstanceResourceDefinition>())
```
It'll throw an `InvalidCastException`. We could add a check to make sure all the objects are the right type, and return a validation message if not?

The reason I didn't bother was that I figured this can't happen as a result of a mistake in the config, it would have to happen as a result of a bug in the code being introduced (for example us breaking the AutoMapper mappings). In that case it would just bubble up and give us a stack trace I think.

What's your thoughts?

tomkerkhove replied:
I think you are right, we are good to go here!

adamconnelly commented on src/Promitor.Scraper.Host/Validation/Steps/MetricsDeclarationValidationStep.cs:
```diff
@@ -77,7 +80,7 @@ private static IEnumerable<string> ValidateMetricDefaults(MetricDefaults metricD
         private static IEnumerable<string> DetectDuplicateMetrics(List<MetricDefinition> metrics)
         {
-            var duplicateMetricNames = metrics.GroupBy(metric => metric.Name)
+            var duplicateMetricNames = metrics.GroupBy(metric => metric.PrometheusMetricDefinition.Name)
```
I'll add null propagation to this. In that case the `MetricsValidator` will also catch it and let us know that no metric name has been defined.

adamconnelly commented on src/Promitor.Scraper.Tests.Unit/Serialization/v1/MetricsDeclaration/YamlSerializationTests.cs:
```diff
@@ -24,16 +25,16 @@ public YamlSerializationTests()
             Assert.NotNull(deserializedMetricDefinition);
-            Assert.Equal(metricDefinition.Name, deserializedMetricDefinition.Name);
-            Assert.Equal(metricDefinition.Description, deserializedMetricDefinition.Description);
+            Assert.Equal(metricDefinition.Name, deserializedMetricDefinition.PrometheusMetricDefinition.Name);
```
Sure.

adamconnelly commented on src/Promitor.Scraper.Tests.Unit/Serialization/v1/MetricsDeclaration/YamlSerializationTests.cs:
```diff
@@ -24,16 +25,16 @@ public YamlSerializationTests()
-            Assert.Equal(deserializedMetricDefinition.ResourceGroupName, metricDefinition.ResourceGroupName);
+            Assert.NotNull(deserializedMetricDefinition.PrometheusMetricDefinition.Labels);
+            Assert.Equal(metricDefinition.Labels, deserializedMetricDefinition.PrometheusMetricDefinition.Labels);
+            Assert.Equal(deserializedMetricDefinition.Resources.Single().ResourceGroupName, metricDefinition.ResourceGroupName);
```
Sure - I've added a not empty assertion since the next line is using `Single()`.

adamconnelly commented on src/Promitor.Scraper.Tests.Unit/Serialization/v1/Mapping/MetricDefinitionV1MappingTests.cs:
```diff
@@ -0,0 +1,63 @@
+            var definition = _mapper.Map<MetricDefinition>(resource);
+
+            // Assert
+            Assert.Equal(resource.Name, definition.PrometheusMetricDefinition.Name);
```
Done.

adamconnelly commented on src/Promitor.Core.Scraping/Scraper.cs:
```diff
@@ -58,28 +58,35 @@ protected Scraper(ScraperConfiguration scraperConfiguration)
-                var scrapedMetricResult = await ScrapeResourceAsync(AzureMetadata.SubscriptionId, resourceGroupName, castedMetricDefinition, aggregationType, aggregationInterval.Value);
+                var aggregationInterval = scrapeDefinition.AzureMetricConfiguration.Aggregation.Interval;
+                var aggregationType = scrapeDefinition.AzureMetricConfiguration.Aggregation.Type;
+                var resourceGroupName = string.IsNullOrEmpty(scrapeDefinition.Resource.ResourceGroupName) ? AzureMetadata.ResourceGroupName : scrapeDefinition.Resource.ResourceGroupName;
```
That's a good point - we've got all the info we need when we create the Scraper, so what I've done is added an extra property to `ScrapeDefinition` that contains the "resolved" resource group name, and removed this code from the scraper. I've also added some tests to make sure we use the correct resource group.

tomkerkhove replied:
Sweet, thanks!

adamconnelly commented:
@tomkerkhove I've addressed most of your comments, and pushed another commit with fixes for them. I've left two unresolved because I'm not completely sure what we should do about them.

Also - I noticed that the Open-Api.xml file changed in the commit I've just pushed. I included it in the commit because I assumed it was just automatically generated based on the code, but I just wanted to point that out because I'm not quite sure how that works.

adamconnelly commented:
Also - just pushed another commit to fix the build. I've just added them as fixups because I figured if you were going to squash the commits before merging anyway it didn't really matter if they had decent commit messages or not.

tomkerkhove commented:
> Also - I noticed that the Open-Api.xml file changed in the commit I've just pushed. I included it in the commit because I assumed it was just automatically generated based on the code, but I just wanted to point that out because I'm not quite sure how that works.

That is correct, you can ignore that one!

Will go through it all now but first glance looks good!
